### ISSUE TITLE
0.17 doesn't seem to use ~/.bitcoin/testnet3 anymore

### ISSUE THREAD

**Issue by zciendor** [bitcoin/bitcoin#14557]
Up to 0.16 it was sufficient to only specify the `-testnet` command line option causing bitcoind to write data to the `~/.bitcoind/testnet3` directory by default and not messing with the mainnet directory `~/.bitcoin`. However, with 0.17 and `-testnet` seems to use the old mainnet directory `~/.bitcoin`.

Not sure if this intended or a bug. Do you have to explicitely specify a different datadir for TESTNET now, if you want to run TESTNET and MAINNET on the same machine?

```
$ /usr/local/bin/bitcoind -conf=/etc/bitcoin/test_bitcoin.conf -testnet
Bitcoin Core version v0.17.0 (release build)
InitParameterInteraction: parameter interaction: -proxy set -> setting -discover=0
InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1
 
************************
EXCEPTION: N5boost10filesystem16filesystem_errorE       
boost::filesystem::create_directory: Permission denied: "/mnt/data/bitcoind/.bitcoin/blocks"       
bitcoin in AppInit()       



************************
EXCEPTION: N5boost10filesystem16filesystem_errorE       
boost::filesystem::create_directory: Permission denied: "/mnt/data/bitcoind/.bitcoin/blocks"       
bitcoin in AppInit()       

Shutdown: In progress...
Shutdown: done
```

ARM-32 (Armbian)

**Comment by MarcoFalke** [comment]
For completeness, would you mind sharing the `/etc/bitcoin/test_bitcoin.conf`?

**Comment by zciendor** [comment]
cat /etc/bitcoin/test_bitcoin.conf
```
# Generated with the help of https://jlopp.github.io/bitcoin-core-config-generator/

# This config should be placed in following path:
# /etc/bitcoin/test_bitcoin.conf

# [core]
# Specify a non-default location to store blockchain and other data.
#datadir=/mnt/data/test_bitcoind/.bitcoin/
# Set database cache size in megabytes; machines sync faster with a larger cache. Recommend setting as high as possible based upon machine's available RAM.
#dbcache=300
# Keep at most <n> unconnectable transactions in memory.
maxorphantx=10
# Keep the transaction memory pool below <n> megabytes.
maxmempool=50
# Set the number of script verification threads. (1 to CPU_CORES, 0 = automatic, less than 0 = leave that many cores free).
# Note: zHIVErbox limits the number of cores used by bitcoind to the 3 BIG cores via /etc/cron.d/make_dapp_processes_faster
#par=-4
# Maintain a full transaction index, used by the getrawtransaction rpc call.
txindex=1

# [debug]
debug=tor
testnet=1

# [network]
# see https://github.com/bitcoin/bitcoin/blob/master/doc/tor.md
# Accept incoming connections from peers.
listen=1
# Tor control <ip:port> to use if onion listening enabled.
torcontrol=127.0.0.1:9052
# connect to peers via Tor.
onlynet=onion 
# SOCKS5 proxy <ip:port> for 'onlynet=onion' configuration.
onion=127.0.0.1:9053
# Connect to other IPv4 and IPv6 peers through <ip:port> SOCKS5 proxy.
proxy=127.0.0.1:9053
# Specify your own public IP address. Only needed for non-ephemeral hidden service
#externalip=.onion
# Allow at most N inbound connections from peers.
maxconnections=48
# Tries to keep outbound traffic under the given target (in MiB per 24h), 0 = no limit.
maxuploadtarget=5000
# Disable UPnP. This was the result of a vulnerability in previous versions of Bitcoin Core and it is suspected that additional vulnerabilities may remain. However, disabling this means that you will certainly need to manually allow incoming connections if you want them.
upnp=0
# Add a node IP address to connect to and attempt to keep the connection open. This option can be set multiple times.
# Since zHIVErbox comes with Cjdns, you can add a few Cjdns adresses of trusted friends here if you like
# However, this won't work with the 'onlynet=onion' option
#addnode=fc00::
#addnode=fc00::

# [relay]
# Fee rates (in BTC/kB) smaller than this are considered zero fee for relaying, mining and transaction creation.
minrelaytxfee=0.0001

# [rpc]
# Location of the RPC auth cookie
# rpccookiefile=/run/test_bitcoind/
# Accept command line and JSON-RPC commands.
# server=1
# Accept public REST requests.
# rest=1
# Username for JSON-RPC connections
# rpcuser=
# Password for JSON-RPC connections
# rpcpassword=
# Allow JSON-RPC connections from specified source. Valid for <ip> are a single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0) or a network/CIDR (e.g. 1.2.3.4/24). This option can be specified multiple times.
# rpcallowip=127.0.0.1
```

**Comment by zciendor** [comment]
The strange thing is that no matter which path I specify (even non-existant) via the `-conf` argument, I get the same error:

$ /usr/local/bin/bitcoind -conf=**/this/is/a/nonexistant/file** -testnet
```
Bitcoin Core version v0.17.0 (release build)
InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1


************************
EXCEPTION: N5boost10filesystem16filesystem_errorE       
boost::filesystem::create_directory: Permission denied: "/mnt/data/bitcoind/.bitcoin/blocks"       
bitcoin in AppInit()       



************************
EXCEPTION: N5boost10filesystem16filesystem_errorE       
boost::filesystem::create_directory: Permission denied: "/mnt/data/bitcoind/.bitcoin/blocks"       
bitcoin in AppInit()       

Shutdown: In progress...
Shutdown: done
```

**Comment by MarcoFalke** [comment]
What datadir and conf file is used according to the debug output of bitcoind?

**Comment by zciendor** [comment]
> What datadir and conf file is used according to the debug output of bitcoind?

The datadir is not specified explicitely anywhere. So it must be the default one within the user's home directory who runs the daemon:
```
test_bitcoind:x:115:999::/mnt/data/bitcoind:/bin/false
```
And the conf file is exactly the one pasted above.

What is special is, that the user `test_bitcoind` doesn't have write permissions to the `/mnt/data/bitcoind/.bitcoin` directory, but it has for the `/mnt/data/bitcoind/.bitcoin/testnet3` subdirectory, which was sufficient until 0.16 when the `-testnet` argument was given.

So the question is, is `~/.bitcoin/testnet3` no longer the default datadir with 0.17 when the `-testnet` argument ist given?

**Comment by promag** [comment]
> The strange thing is that no matter which path I specify (even non-existant) via the `-conf` argument, I get the same error:
> 
> $ /usr/local/bin/bitcoind -conf=**/this/is/a/nonexistant/file** -testnet

I don't get the error with current master (613fc95), instead I get the warning:
```
Warning: The specified config file /this/is/a/nonexistant/file does not exist
```

**Comment by zciendor** [comment]
@jnewbery

**Comment by MarcoFalke** [comment]
Moving to 0.17.2

**Comment by MarcoFalke** [comment]
Is this fixed by #14409? @zciendor

**Comment by benthecarman** [comment]
My setup is running up-to-date with master and it uses `~/.bitcoin/testnet3` for my testnet data

**Comment by MarcoFalke** [comment]
Is this still an issue on current master?

**Comment by zciendor** [comment]
> Is this fixed by #14409? @zciendor

I can confirm this is fixed with release 0.17.1
